File: magmac/src/java/magma/compile/rule/TypeRule.java
package magma.compile.rule;

import magma.api.result.Err;
import magma.api.result.Result;
import magma.compile.CompileError;
import magma.compile.CompileParentError;
import magma.compile.Error_;
import magma.compile.rule.result.ErrorRuleResult;
import magma.compile.rule.result.RuleResult;
import magma.java.JavaOptionals;

public record TypeRule(String type, Rule child) implements Rule {

    public static final String FORMAT = "Node was not of type '%s', but rather '%s'.";

    @Override
    public RuleResult toNode(String input) {
        var result = child.toNode(input);
        if (JavaOptionals.toNative(result.findError()).isEmpty()) return result.withType(type);

        var format = "Cannot attach type '%s' because of child failure.";
        var message = format.formatted(type);
        return new ErrorRuleResult(new CompileParentError(message, input, JavaOptionals.toNative(result.findError()).get()));
    }

    @Override
    public Result<String, Error_> fromNode(Node node) {
        if (!node.is(type)) {
            var message = String.format(FORMAT, type, node.findType());
            return new Err<>(new CompileError(message, node.toString()));
        }

        return child.fromNode(node).mapErr(err -> {
            var format = "Cannot generate '%s' from node.";
            var message = format.formatted(type);
            return new CompileParentError(message, node.toString(), err);
        });
    }
}

File: src/clang/magma/compile/rule/TypeRule.c
package magma.compile.rule;class package magma.compile.rule;{package magma.compile.rule;

import magma.compile.CompileError;class import magma.compile.CompileError;{

import magma.compile.CompileError;
import magma.compile.Node;class import magma.compile.Node;{
import magma.compile.Node;
import magma.compile.context.NodeContext;class import magma.compile.context.NodeContext;{
import magma.compile.context.NodeContext;
import magma.result.Err;class import magma.result.Err;{
import magma.result.Err;
import magma.result.Result;class import magma.result.Result;{
import magma.result.Result;

public record TypeRule(String type, Rule rule) implements Rule {
    @Override
    public Result<Node, CompileError> parse(String input) {
        return rule.parse(input).mapValue(node -> node.withType(type));
    }

    @Override
    public Result<String, CompileError> generate(Node node) {
        if (node.is(type)) {
            return rule.generate(node);
        }

        return new Err<>(new CompileError("Node was not of type '" + type + "'", new NodeContext(node)));
    }
}
class public record TypeRule(String type, Rule rule) implements Rule {
    @Override
    public Result<Node, CompileError> parse(String input) {
        return rule.parse(input).mapValue(node -> node.withType(type));
    }

    @Override
    public Result<String, CompileError> generate(Node node) {
        if (node.is(type)) {
            return rule.generate(node);
        }

        return new Err<>(new CompileError("Node was not of type '" + type + "'", new NodeContext(node)));
    }
}{

public record TypeRule(String type, Rule rule) implements Rule {
    @Override
    public Result<Node, CompileError> parse(String input) {
        return rule.parse(input).mapValue(node -> node.withType(type));
    }

    @Override
    public Result<String, CompileError> generate(Node node) {
        if (node.is(type)) {
            return rule.generate(node);
        }

        return new Err<>(new CompileError("Node was not of type '" + type + "'", new NodeContext(node)));
    }
}
